packages/matter.js/src/cluster/ClusterServerFactory.ts

```diff
@@ -0,0 +1,122 @@
+/**
+ * @license
+ * Copyright 2022-2023 Project CHIP Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ClusterServer } from "../protocol/interaction/index.js";
+import { BitSchema, TypeFromBitSchema } from "../schema/BitmapSchema.js";
+import { Attributes, Cluster, Commands, Events, GlobalAttributes, MandatoryLocalAttributeNames, OptionalAttributeNames } from "./Cluster.js";
+import { AttributeInitialValues, ClusterServerHandlers } from "./server/ClusterServer.js";
+import { UnsupportedCommandError } from "./server/CommandServer.js";
+
+type AnyCluster = Cluster<any, any, any, any, any>;
+const GLOBAL_ATTRIBUTES = new Set(Object.keys(GlobalAttributes({})));
+
+export function notImplemented() {
+    // Command server will fail gracefully with the appropriate status code
+    // in response to this error
+    throw new UnsupportedCommandError();
+}
+
+function mandatoryLocalAttributeNames<A extends Attributes>(attributes: A) {
+    return Object.entries(attributes)
+        .filter(([_name, attr]) => !attr.optional)
+        .filter(([name]) => !GLOBAL_ATTRIBUTES.has(name))
+        .map(([name]) => name as MandatoryLocalAttributeNames<A>);
+}
+
+function optionalAttributeNames<A extends Attributes>(attributes: A) {
+    return Object.entries(attributes)
+        .filter(([_name, attr]) => attr.optional)
+        .map(([name]) => name as OptionalAttributeNames<A>)
+}
+
+// Shortcut for cluster server implementers - create initial attribute map from
+// schema if not provided.  Could move into ClusterServer but then
+// attributeInitialValues argument would need to be optional.  Not sure if
+// that's desirable.  Also, this file is shorter than InteractionServer ;)
+function createDefaultAttributeInitialValues<F extends BitSchema, SF extends TypeFromBitSchema<F>, A extends Attributes, C extends Commands, E extends Events>(cluster: Cluster<F, SF, A, C, E>): AttributeInitialValues<A> {
+    const result = {} as AttributeInitialValues<A>;
+
+    mandatoryLocalAttributeNames(cluster.attributes).forEach((name) =>
+        (result as any)[name] = cluster.attributes[name].default);
+
+    // TODO - caller can override but should we make the set of optional
+    // attributes configurable?
+    optionalAttributeNames(cluster.attributes).forEach((name) => {
```

Comment:
Idea for optional is that the dev decides if he supports them. So: if he sets a value he states support of them. If not then not. Means: we should not populate default values for optionales at all. 

Ps: we also can not be sure that all mandatory attributes have a default value ... just saying.

Reply:
OK, that's how I had it originally but the only optional value with a default was the lastConfiguredBy: null in ScenesCluster so I wasn't sure if default for options was intended to designate they're supported by default.

If you look at registerClusterDefaults, I added the defaultAttributeInitialValues there so the server implementation could override.  Do you think it would be OK to make it a partial so they don't have to copy all of the default values and could just change the ones that diverge from the schema?

Reply:
When we do Partial we need to throw if a value is missing... On the other hand - also because some values depend on each other it might make smore sense to say" you want to define, define all" because else dev also need to check the cluster def where we have a defautl and where mot and such ... will in the end become more complex and effort for the dev with an Partial.

So my conclusion would be: lets do with "full or nothing" (kind of) ... We can change to Partial any time later if we want - and we need to throw also anyweay ... sooo

Reply:
So the first use case is ScenesClusterServer.  Either has to copy 5 default values out of the schema or just add `lastConfiguredBy: null`.  Not sure if that's a representative use case.

Other 3 are fine though with just mandatory defaults.

Reply:
ScenesCluster is easy to fix ... see cluster def: "Null" is default - and in the end is optional ... so we can also simply skip it until we support it)